### ISSUE TITLE
pre-trained models  for GRCh38/hg38

### ISSUE THREAD

**Issue by ulderma** [timothyjamesbecker/FusorSV#30]
Hello Timothy,

Thank you for developing FusorSV. I'm trying to use it with GRCh38/hg38 data.

Could you please advise:
1. Are there any pre-trained models available for GRCh38/hg38?
2. If not, what training data (truth set) would you recommend for training a model on GRCh38? (e.g., specific GIAB samples, file formats, etc.)
3. Are there any plans to release GRCh38-compatible models in the future?

Thank you for your time and guidance.

**Comment by timothyjamesbecker** [comment]
Hello, you can use FusorSV with GRCh38/hg38 workflows (IE use SV callers on GRCh38/hg38) with the base model which is human_g1k_v37_decoy.fa based (hg19 variant).  If you are passing a new reference into the main FusorSV.py script such as Hg38.fa, it should build out the necessary coordinate map JSON files for you. If you are having trouble you can try to post some of your command line here as well as any errors you may receive. 

It is not recommended to use this model with other non-mammalian genomes as the SV callers could be not working as well or in the same ways.

**Comment by ulderma** [comment]
Hello,

I'm writing to you regarding FusorSV. When running the pipeline to merge structural variants (SVs) from two VCF files (from Manta and Delly), the script runs but doesn't generate the expected merged VCF file with the results. Instead, only auxiliary files and a log are created.


**Input data:**

Manta VCF: sample1_S48.vcf (11,713 rows with variants)

Delly VCF: sample1_S11.vcf (32,512 rows with variants)

Examples of the first lines of each file are attached below.


**FusorSV version:**
(sv-pipeline) anna@ccu-dev4:~/anna/ulderma$ echo "Manta:" $(grep -vc "^#" ~/anna/ulderma/vcfs_for_fusor_simple/sample1/sample1_S48.vcf)
Manta: 11713
(sv-pipeline) anna@ccu-dev4:~/anna/ulderma$ echo "Delly:" $(grep -vc "^#" ~/anna/ulderma/vcfs_for_fusor_simple/sample1/sample1_S11.vcf)
Delly: 32512
(sv-pipeline) anna@ccu-dev4:~/anna/ulderma$ python /home/anna/miniconda3/envs/sv-pipeline/bin/FusorSV.py --version 2>&1 | head -5
usage: FusorSV.py [-h] [-r REF_PATH] [-i IN_DIR] [-a CTG_DIR] [-c CHROMS]
                  [-o OUT_DIR] [-m SV_MASK] [-f APPLY_FUSION_MODEL_PATH]
                  [-p CPUS] [--k_fold K_FOLD] [--n_k N_K] [--bins BINS]
                  [--obs OBS] [--min_g MIN_G] [--over_m OVER_M]
                  [--pre_cluster] [--smoothing] [--detail]



**Startup script:**
python /home/anna/miniconda3/envs/sv-pipeline/bin/FusorSV.py \
  -i ~/anna/ulderma/vcfs_for_fusor_simple/sample1/ \
  -r /home/anna/ngs-data/ngs/ref/hg38.fa \
  -m ~/anna/ulderma/fusorsv_input/hg38.bed \
  -o fusor_hg38_new \
  --pre_cluster \
  --bins 1000 \
  --min_g 0.1 2>&1 | tee fusor_log.txt



**fusor_log.txt:**
(sv-pipeline) anna@ccu-dev4:~/anna/ulderma$ cat fusor_log.txt
no contig directory specified
processing samples ['/home/anna/anna/ulderma/vcfs_for_fusor_simple/sample1/sample1_S48.vcf', '/home/anna/anna/ulderma/vcfs_for_fusor_simple/sample1/sample1_S11.vcf']
 for chroms ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y', 'MT']
making a new coordinate offset json file
merging the svmask regions
svmask regions merged in 0.0 sec
reading, parsing, partitioning and writing sample VCFs
reading sample sample1_S48.vcf
reading sample sample1_S11.vcf
finished reading 2 out of 2 samples generating 0 partitions in 0.55 sec
starting model partition:       t=0     b=0
starting model partition:       t=0     b=1

****************

starting model partition:       t=7     b=7
starting model partition:       t=7     b=8
finished modeling in 38.04 sec
apply fusion model to sample inputs and generating fusorSV ouput
starting fusorSV discovery on sample sample1_S48.vcf
starting fusorSV discovery on sample sample1_S11.vcf
finished reading samples in 26.33 sec
run 0 in 74.45 sec
:::::::::::::::::::::


Problem: After this line, execution terminates. The resulting VCF file is not created in the output directory (fusor_hg38_new). Only the contigs_offset.json and merged_svmask.bed files and the partition_files folder are created.

**Comment by ulderma** [comment]
**Here are 2 vcf files:**
(sv-pipeline) anna@ccu-dev4:~/anna/ulderma$ grep -v "^#" ~/anna/ulderma/vcfs_for_fusor_simple/sample1/sample1_S48.vcf | head -5
chr1    839442  MantaDEL:25:0:0:1:0:0   CACCTAGACACACACTCCTGGACACACACACCTAGACACACACACCTGGACAAACACA      CT      180     PASS    END=839499;SVTYPE=DEL;SVLEN=-57;CIGAR=1M1I57D   GT:FT:GQ:PL:PR:SR       0/1:PASS:53:230,0,50:3,0:6,8
chr1    998763  MantaINS:1:15:15:0:0:0  GC      GGGGAGGGCGCGGAGCGGAGGGGAGGGCGCGGAGCGGAGGGGAGGGCGCGGAGCGGAGG     922     PASS    END=998764;SVTYPE=INS;SVLEN=58;CIGAR=1M58I1D    GT:FT:GQ:PL:PR:SR       1/1:PASS:51:975,54,0:0,0:11,24
chr1    1068824 MantaINS:1:24:24:0:0:0  G       GGCCACGCGGGCTGTGCAGATGCAGGTGCGGCGGGGCGGGGCCACGCGGGCTGTGAAGGTGCAGGTGCGGCGGGGCAGA 372     PASS    END=1068824;SVTYPE=INS;SVLEN=78;CIGAR=1M78I;CIPOS=0,10;HOMLEN=10;HOMSEQ=GCCACGCGGG    GT:FT:GQ:PL:PR:SR       0/1:PASS:48:422,0,45:0,0:8,20
chr1    1288378 MantaDEL:159:0:1:0:0:0  C       <DEL>   209     PASS    END=1290055;SVTYPE=DEL;SVLEN=-1677;IMPRECISE;CIPOS=-293,293;CIEND=-398,398      GT:FT:GQ:PL:PR  0/1:PASS:209:259,0,255:19,15
chr1    1350107 MantaDEL:1:39:41:0:0:0  G       <DEL>   216     SampleFT        END=1351389;SVTYPE=DEL;SVLEN=-1282;IMPRECISE;CIPOS=-319,319;CIEND=-393,394      GT:FT:GQ:PL:PR  0/1:MinGQ:11:265,0,7:3,13

and


(sv-pipeline) anna@ccu-dev4:~/anna/ulderma$ grep -v "^#" ~/anna/ulderma/vcfs_for_fusor_simple/sample1/sample1_S11.vcf | head -5
chr1    10326   DEL00000001     C       N       2       LowQual IMPRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv0.9.1;END=180869;PE=2;MAPQ=1;CT=3to5;CIPOS=-829,829;CIEND=-829,829     GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV       0/0:0,-0.549364,-2.7473:6:LowQual:146:7156:2413:6:2:0:0:0
chr1    611464  DUP00000005     C       N       15      LowQual IMPRECISE;SVTYPE=DUP;SVMETHOD=EMBL.DELLYv0.9.1;END=611922;PE=2;MAPQ=10;CT=5to3;CIPOS=-595,595;CIEND=-595,595    GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV       0/0:0,-5.86598,-62.35:59:PASS:11684:19581:851:3:24:2:0:0
chr1    651032  DEL00000006     C       N       66      PASS    PRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv0.9.1;END=651059;PE=0;MAPQ=0;CT=3to5;CIPOS=-20,20;CIEND=-20,20;SRMAPQ=22;INSLEN=0;HOMLEN=20;SR=3;SRQ=1;CONSENSUS=TCAGGAAGACAAGTATGAAGCATAATCTGTGACATTATTGATATCTTCCTGAAGAACATAATTCCTGCCTACCATCAACAAGCATCAATACTTTCTACCA;CE=1.91515       GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV 0/1:-6.94605,0,-31.4949:69:PASS:267:515:358:2:0:0:15:6
chr1    789469  DEL00000007     C       N       25      LowQual PRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv0.9.1;END=790145;PE=0;MAPQ=0;CT=3to5;CIPOS=-21,21;CIEND=-21,21;SRMAPQ=4;INSLEN=0;HOMLEN=27;SR=3;SRQ=0.965035;CONSENSUS=ATGGAATGCAATGTAATGGACTCGAATGGAACGGAATGGAATGGACAAGAATTGAATTGAATGGACTGGAATGGAATGGAATGGAATGCAATGGAATGCACTCGAACGGATTGGAATGGAATGGACTCGAATAGAATGGAAT;CE=1.84545       GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV       0/1:-52.1689,0,-36.0896:10000:PASS:64:88:125:1:0:0:16:22
chr1    820901  DEL00000013     G       N       64      LowQual IMPRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv0.9.1;END=823478;PE=3;MAPQ=18;CT=3to5;CIPOS=-362,362;CIEND=-362,362    GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV       0/0:0,-5.93905,-202.132:59:PASS:204:479:249:2:38:3:0:0



I need a result similar to what my weighted voting script generates (see screenshot voting_results.vcf), where for each PASS option the following are specified:

SVTYPE (option type)

WEIGHT (weight/support)

CALLERS (which tools detected it)


I'm attaching a screenshot below

<img width="2050" height="886" alt="Image" src="https://github.com/user-attachments/assets/bdbc23f2-2459-4034-9bd3-4e9e80cdefc7" />

**Comment by ulderma** [comment]
**Questions:**

Is the above command correct for generating a consensus VCF?

Based on the log, why didn't the script create the output file? Could the --bins, --min_g parameters, or the missing -f (APPLY_FUSION_MODEL_PATH) argument have caused the script to exit silently without errors but without any output?
 
Do VCFs from Manta and Delly need to be specially prepared or filtered in any way (for example, by leaving only the PASS variants or converting the INFO formats)? Currently, they are used as is.

Thank you in advance for your help with this issue. I am happy to provide any further details if needed.

Kind regards,
Ulviya

**Comment by ulderma** [comment]
The difference is only 1 bp at END! But FusorSV didn't merge it.

Manta: chr8:125324555-125324872 (DEL, 317 bp)
Delly: chr8:125324555-125324873 (DEL, 318 bp)

Example:

(base) anna@ccu-dev4:~/anna/ulderma$ grep -v "^#" ~/anna/ulderma/vcfs_for_fusor_simple/sample1/sample1_S11.vcf | head -5
chr1    10326   DEL00000001     C       N       2       LowQual IMPRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv0.9.1;END=180869;PE=2;MAPQ=1;CT=3to5;CIPOS=-829,829;CIEND=-829,829     GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV       0/0:0,-0.549364,-2.7473:6:LowQual:146:7156:2413:6:2:0:0:0
chr1    611464  DUP00000005     C       N       15      LowQual IMPRECISE;SVTYPE=DUP;SVMETHOD=EMBL.DELLYv0.9.1;END=611922;PE=2;MAPQ=10;CT=5to3;CIPOS=-595,595;CIEND=-595,595    GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV       0/0:0,-5.86598,-62.35:59:PASS:11684:19581:851:3:24:2:0:0
chr1    651032  DEL00000006     C       N       66      PASS    PRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv0.9.1;END=651059;PE=0;MAPQ=0;CT=3to5;CIPOS=-20,20;CIEND=-20,20;SRMAPQ=22;INSLEN=0;HOMLEN=20;SR=3;SRQ=1;CONSENSUS=TCAGGAAGACAAGTATGAAGCATAATCTGTGACATTATTGATATCTTCCTGAAGAACATAATTCCTGCCTACCATCAACAAGCATCAATACTTTCTACCA;CE=1.91515       GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV 0/1:-6.94605,0,-31.4949:69:PASS:267:515:358:2:0:0:15:6
chr1    789469  DEL00000007     C       N       25      LowQual PRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv0.9.1;END=790145;PE=0;MAPQ=0;CT=3to5;CIPOS=-21,21;CIEND=-21,21;SRMAPQ=4;INSLEN=0;HOMLEN=27;SR=3;SRQ=0.965035;CONSENSUS=ATGGAATGCAATGTAATGGACTCGAATGGAACGGAATGGAATGGACAAGAATTGAATTGAATGGACTGGAATGGAATGGAATGGAATGCAATGGAATGCACTCGAACGGATTGGAATGGAATGGACTCGAATAGAATGGAAT;CE=1.84545       GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV       0/1:-52.1689,0,-36.0896:10000:PASS:64:88:125:1:0:0:16:22
chr1    820901  DEL00000013     G       N       64      LowQual IMPRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv0.9.1;END=823478;PE=3;MAPQ=18;CT=3to5;CIPOS=-362,362;CIEND=-362,362    GT:GL:GQ:FT:RCL:RC:RCR:RDCN:DR:DV:RR:RV       0/0:0,-5.93905,-202.132:59:PASS:204:479:249:2:38:3:0:0
(base) anna@ccu-dev4:~/anna/ulderma$ grep "125324555" ~/anna/ulderma/vcfs_for_fusor_simple/sample1/sample1_S48.vcf
chr8    125324555       MantaDEL:4958:6:6:4:0:0 TCCATTTTGATAAAATTCTTTTTTTTTTTTTTTTTTTTTGAGACGAGTCTGGCTCTGTTGCCCAGGCTGGAGTGCAGTGGCGCAATCTCGGCTCACTGCAAGCTCCGCCTCCCGGGTTCACGCCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCGCCCGCCACTACGCCCGGCTAATTTTTTGTATTTTTAGTAGAGACAGGGTTTCACCATGTTAGACAGGATGGTCTCGATCTCCTGACCTCGTGATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAGGCATGAGCCACCGCGCCCGGC        T       186     PASS    END=125324872;SVTYPE=DEL;SVLEN=-317;CIGAR=1M317D;CIPOS=0,20;HOMLEN=20;HOMSEQ=CCATTTTGATAAAATTCTTT     GT:FT:GQ:PL:PR:SR       0/1:PASS:185:236,0,188:34,22:22,9

The variants are almost identical, but FusorSV doesn't merge" — example chr8:125324555

Many different types of SV" — examples from Delly (large deletions vs. short ones in Manta)

My script found 2,712 intersections, FusorSV found 0.


Why doesn't FusorSV combine variants with a 1 bp difference in END? Should CIPOS/CIEND be taken into account when comparing?

**Comment by timothyjamesbecker** [comment]
Commenting on the command you are using. This is user error: you should read the paper and the documents to see that FusorSV takes in a folder of sample folders where each sample folder has at least 2-callers of each class (2 read depth like CNVnator/CNVpytor, 2 paired-end like Lumpy/Breakdancer, 2 assembly like Delly/Tiagra) The paper details the caller ensemble that should be used with the based model. If you are simply merging 2 callers such as Manta and CNVpytor than you can use bedtools2 and you don't need any scripts! FusorSV uses machine learning to find the optimal merging strategies through a data fusion structure as the methods in the paper detail fully.  

Confidence intervals are not used since they are not build from robust external distributions but are simply estimates from the sample read insertion statistics and are not meaningful when used in an ensemble system which instead measures end-point inference accuracy.

If you are not training your own model you have to use a preexisting fusion  via the -f parameter as the readme explains. This is not optional and is how you are supposed to use this method.

Closing due to a lack of issue (IE read the documentation to see how the method should be used)

**Comment by ulderma** [comment]
Sorry for my stupid questions, I'm not a geneticist or bioinformatician. I thought I could check two VCFs and then upload five or more. Sorry, thanks for the answers.

**Comment by timothyjamesbecker** [comment]
You can check it with only one file, but you will still have to set it up as a folder of sample folders as the docs show:

/full_experiment_folder/each_sample_folder/sample1_S48.vcf #this one in Manta

You can also download the example folders that already have this structure which can be used to build the model it is is the very first release here: 

https://github.com/timothyjamesbecker/FusorSV/releases/download/0.1.0-beta/meta_caller_RC1.tar.gz

This can be used to make sure that your FusorSV.py build is correct since I see you are not using the docker version. It can also be used to see how to structure your files.

If you want to do a minimal version of this I also made a minimal caller model in the newest release here:

https://github.com/timothyjamesbecker/FusorSV/releases/tag/0.1.4

If you instead want to simply merge two callers (and not filter or remove calls) you can just use bedtools2 intersection command line tool:

http://quinlanlab.org/tutorials/bedtools.html

There are many other tools that can also do that and you should do a search at biostars.com where bioinformatics folks can help!  Good luck with your work.
